mypy error on progress bar usage

xxx.py:
```python
import click

with click.progressbar([1, 2, 3]) as bar:
    for x in bar:
        pass
```

mypy error:
```
$ mypy xxx.py 
xxx.py:4: error: "Iterable[<nothing>]" has no attribute "__next__"  [attr-defined]
Found 1 error in 1 file (checked 1 source file)
```

Environment:

- Python version: 3.9.5
- Click version: 8.0.0
- mypy version: 0.812

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -10,6 +10,8 @@ Unreleased
 -   Annotate ``Context.obj`` as ``Any`` so type checking allows all
     operations on the arbitrary object. :issue:`1885`
 -   Fix some types that weren't available in Python 3.6.0. :issue:`1882`
+-   Fix type checking for iterating over ``ProgressBar`` object.
+    :issue:`1892`
 
 
 Version 8.0.0

--- a/src/click/_termui_impl.py
+++ b/src/click/_termui_impl.py
@@ -32,7 +32,7 @@ else:
     AFTER_BAR = "\033[?25h\n"
 
 
-class ProgressBar:
+class ProgressBar(t.Generic[V]):
     def __init__(
         self,
         iterable: t.Optional[t.Iterable[V]],
@@ -50,7 +50,7 @@ class ProgressBar:
         color: t.Optional[bool] = None,
         update_min_steps: int = 1,
         width: int = 30,
-    ):
+    ) -> None:
         self.fill_char = fill_char
         self.empty_char = empty_char
         self.bar_template = bar_template
@@ -101,7 +101,7 @@ class ProgressBar:
     def __exit__(self, exc_type, exc_value, tb):  # type: ignore
         self.render_finish()
 
-    def __iter__(self) -> t.Iterable[V]:
+    def __iter__(self) -> t.Iterator[V]:
         if not self.entered:
             raise RuntimeError("You need to use progress bars in a with block.")
         self.render_progress()
@@ -113,7 +113,7 @@ class ProgressBar:
         # because `self.iter` is an iterable consumed by that generator,
         # so it is re-entry safe. Calling `next(self.generator())`
         # twice works and does "what you want".
-        return next(iter(self))  # type: ignore
+        return next(iter(self))
 
     def render_finish(self) -> None:
         if self.is_hidden:

--- a/src/click/termui.py
+++ b/src/click/termui.py
@@ -318,7 +318,7 @@ def progressbar(
     file: t.Optional[t.TextIO] = None,
     color: t.Optional[bool] = None,
     update_min_steps: int = 1,
-) -> "ProgressBar":
+) -> "ProgressBar[V]":
     """This function creates an iterable context manager that can be used
     to iterate over something while showing a progress bar.  It will
     either iterate over the `iterable` or `length` items (that are counted

